### ISSUE TITLE
SubscriptionListener doesn't inform about first entities

### ISSUE THREAD

**Issue by Namek** [junkdog/artemis-odb#248]
When I register `EntitySubscription.SubscriptionListener` (through ugly reflected inspection) I get informed only about entities added/removed during the game.

I've got `GameSessionManager extends Manager` which has overriden `initialize()` method. That method adds some entities to the world. The problem is that I don't get informed about those entities added in through that method. @junkdog Is that correct behavior? If it is, then probably any workaround to get information about ALL added/removed entities?

Side note: I'm working on my custom entity tracker (probably something similiar to `dominatrix-odb`)

**Comment by junkdog** [comment]
> (through ugly reflected inspection)

Why do you have to use reflection? It's possible I've left something out of the code, but:

1. Make sure your manager implements `SubscriptionListener`
1. get the `EntitySubscription` from `AspectSubscriptionManager`
1. register your manager: `EntitySubscription#addSubscriptionListener`

The above will/should only inform you about which entities have been added/removed from the `EntitySubscription`. If you want to be informed about all entities, override the EntityObserver stubs in Manager (added, changed, deleted).

Be aware that the EntitySubscription code hasn't been finalized yet, so you're on the extreme end of bleeding edge atm. I figured I'd try to polish it once everything is working; currently resolving a profile weaver bug.

**Comment by Namek** [comment]
`AspectSubscriptionManager` has private things, the only way to get `EntitySubscription` is by having `Aspect.Builder` which is private in `EntitySystem`. So I get `Aspect.Builder` using reflection.

> If you want to be informed about all entities, override the EntityObserver stubs in Manager (added, changed, deleted).

I tried to override any/both of `#inserted` methods in EntitySystem and I think there's a bug because I don't get notified about entities at all. Overriding `added` in `Manager` works, thanks.

I know it's a bleeding edge. I rather wait for your API to be finalized but above (first entities creation is not broadcasted) seems like a bug to me so I decided to ask.

**Comment by junkdog** [comment]
I'll see to it first thing tomorrow; been lazy with test cases... heh.

**Comment by junkdog** [comment]
Could you try it now - it's on the *extract-base-system-from-es* branch? I I think I have it working, unless I've missed an edge case.

**Comment by junkdog** [comment]
> AspectSubscriptionManager has private things, the only way to get EntitySubscription is by having Aspect.Builder which is private in EntitySystem. So I get Aspect.Builder using reflection.

I've locally added a getAspectBuilder method to EntitySubscription. While not maybe what you're looking for, you can pass a new Aspect.Builder to `AspectSubscriptionManager`.

**Comment by junkdog** [comment]
@Namek Do you have an RSS feed to your blog?

**Comment by DaanVanYperen** [comment]
http://www.namekdev.net/?

**Comment by junkdog** [comment]
That was easy, don't know why my rss reader didn't pick it up.

**Comment by Namek** [comment]
> Could you try it now - it's on the extract-base-system-from-es branch? I I think I have it working, unless I've missed an edge case.

Well, ok. First - I like results of #247! But I wonder why `VoidEntitySystem`  does extend `BaseSystem` and not `EnitySystem` (hence the name). This way I can't simply implement `inserted` and `removed` when I have no aspect and still want to have some kind of `update()` method (currently: `processSystem()`). Instead, I can...

> While maybe not what you're looking for, you can pass a new Aspect.Builder to AspectSubscriptionManager in order to retrieve the EntitySubscription instance.

and I did that this way:

```java
AspectSubscriptionManager am = world.getManager(AspectSubscriptionManager.class);
am.get(Aspect.all()).addSubscriptionListener(this);
```

```java
public class EntityTracker extends VoidEntitySystem implements EntitySubscription.SubscriptionListener { // ...
```

and works :) so I'm closing this issue. Thanks!

**Comment by Namek** [comment]
@junkdog Side question: will Managers even exist in 1.0? Those EntitySubscriptions look quite like a replacement.

**Comment by DaanVanYperen** [comment]
> ut I wonder why VoidEntitySystem does extend BaseSystem and not EnitySystem (hence the name).

Javadoc describes as a class just for entityless and aspectless game logic. Your comment makes sense though, there's probably a lot of legacy around this class. 

Why not move it into EntitySystem hierarchy for legacy sake and train people to prefer `BaseSystem` instead (I'm loving `BaseSystem` btw).

**Comment by Namek** [comment]
@DaanVanYperen @junkdog another side question: how to call managers and basesystems by one common name? Previously I could tell that they're entity observers but after those cool `BaseSystem` appeared...

**Comment by junkdog** [comment]
> Why not move it back into EntitySystem hierarchy for legacy sake and train people to prefer BaseSystem instead (I'm loving BaseSystem btw).

Yup, it makes more sense; re-opening the issue until it's fixed. Could just deprecate VoidEntitySystem.


@Namek atm, it isn't possible; it *might* come in the future, but I can't decide if it's adding any value or just breaking code unnecessarily - I go back and forth. The new BaseSystem is so bare-bones that it can serve as a Manager too, except for the missing EntityObserver interface. World#setSystem should probably intercept systems implementing EntityObserver either way.

**Comment by Namek** [comment]
Just reporting - subscriptions still don't work as expected. I use `v0.9.1-SNAPSHOT` right now.

Even if I get notified for `Aspect.all()`, I don't get notified about other aspects, like `Aspect.all(Renderable.class)`.

So even in `EntitySystem` I have to manually register subscription listener:

```java
protected void initialize() {
  AspectSubscriptionManager am = world.getManager(AspectSubscriptionManager.class);
  am.get(Aspect.all()).addSubscriptionListener(this);
}
```

**Comment by Namek** [comment]
Another note:
```java
world.createEntity().edit().create(Renderable.class);
```
When I get notified about this entity, it doesn't have `Renderable` component (`e.getComponent(Renderable.class)` returns `null`). So it would be cool to be informed about new entity in next frame.

**Comment by junkdog** [comment]
Thanks for reporting. Do you know if the bug occurs in 0.9.0 too (alt I can
check tomorrow)?

Hmm, might try to do an early release 0.10.0 beginning of next week
(postponing world serialization until 0.11.0). There are a couple of other
bugs I want resolved too + working subscription listeners would be neat.

**Comment by Namek** [comment]
Ahh, sorry. Did you check it already?

**Comment by Namek** [comment]
I can confirm that subscribing for specific Aspect doesn't get me notified about insertion in `0.9.0`.

I put systems in this order:
1. `System1 extends BaseSystem`
2. `System2 extends BaseSystem implements SubscriptionListener`.

`System1`:
```java
	@Override
	protected void processSystem() {
		if (!isInitialized) {
			world.createEntity().edit().create(MyComponent.class);
			isInitialized = true;
		}
	}
```

Now `System2`:
```java
	@Override
	protected void initialize() {
		AspectSubscriptionManager am = world.getManager(AspectSubscriptionManager.class);
		am.get(Aspect.all(MyComponent.class)).addSubscriptionListener(this);
	}

	@Override
	public void inserted(ImmutableBag<Entity> entities) {
		MyComponent cmp = mapper.get(entities.get(0));
		// breakpoint never gets me here
	}
```

**Comment by junkdog** [comment]
> I can confirm that subscribing for specific Aspect doesn't get me notified about insertion in 0.9.0.

But the above works in 0.9.1-SNAPSHOT, right? If not, how does [https://github.com/junkdog/artemis-odb/blob/3509cfffbb7b0f3859cabc6cbc834f76543272e2/artemis/src/test/java/com/artemis/SubscriptionListenerTest.java](this) differ?

**Comment by Namek** [comment]
AFAIR it doesn't

**Comment by junkdog** [comment]
> MyComponent cmp = mapper.get(entities.get(0));
> // breakpoint never gets me here

Did you add `@Wire` to System2 btw?

**Comment by junkdog** [comment]
Possibly fixed by #268

**Comment by Namek** [comment]
Not sure about `@Wire` then. I'll be testing that in following days. Gonna report if that's still broken.

**Comment by Namek** [comment]
I can't reproduce this bug anymore. I use artemis quickstart and Gradle, even if I put artemisVersion=0.9.0 it still downloads 0.9.1-SNAPSHOT dated at April 8th.

The #268 was just another bug, I think this one here was fixed earlier in 0.9.1-SNAPSHOT.

Anyway, the initial problem seems to be fixed. Thanks :)

**Comment by junkdog** [comment]
Nice, well - I better release 0.10.0 then.
